File: javaver/BFSSolver.java
package javaver;

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

public class BFSSolver extends MazeSolver{
    public BFSSolver(Maze maze, Cell start){
        super(maze, start);
    }
    public BFSSolver(String FileName){
        super(FileName);
    }
    public String[] solve(){
        Queue<Cell> q = new LinkedList<Cell>();
        ArrayList<String> solution = new ArrayList<String>();
        q.add(start);
        File f = new File(".\\solutions\\solution0.txt");
        String s = "";
        while(q.peek() != null){
            if(q.peek().getValue() == 3){
                path.push(q.peek());
                System.out.println("found end");
                break;
            }
            int x = q.peek().getX();
            int y = q.peek().getY();
            //System.out.println("checking " + q.peek());
            s += "checking " + q.peek() + "\n";
            solution.add("checking " + q.peek());
            if(x > 0 && (maze.getCell(x-1, y).getValue()  == 0 || maze.getCell(x-1, y).getValue() == 3)&& maze.getCell(x-1,y).isVisited() == false){
                q.add(maze.getCell(x-1, y));
                maze.getCell(x-1, y).setPrevCell(maze.getCell(x, y));
                maze.getCell(x-1, y).visit();
                // System.out.println("added " + maze.getCell(x-1, y));
                solution.add("added " + maze.getCell(x-1, y));
                s += "added " + maze.getCell(x-1, y) + "\n";
            }
            if (y > 0 && (maze.getCell(x, y-1).getValue() == 0 || maze.getCell(x, y-1).getValue() == 3) && maze.getCell(x,y-1).isVisited() == false){
                q.add(maze.getCell(x, y-1));
                maze.getCell(x, y-1).setPrevCell(maze.getCell(x, y));
                maze.getCell(x, y-1).visit();
                // System.out.println("added " + maze.getCell(x, y-1));
                solution.add("added " + maze.getCell(x, y-1));
                s += "added " + maze.getCell(x, y-1) + "\n";
            }
            if(x < maze.getWidth() - 1 && (maze.getCell(x + 1, y).getValue() == 0 || maze.getCell(x + 1, y).getValue() == 3) && maze.getCell(x+1,y).isVisited() == false){
                q.add(maze.getCell(x + 1, y));
                maze.getCell(x + 1, y).setPrevCell(maze.getCell(x, y));
                maze.getCell(x + 1, y).visit();
                // System.out.println("added " + maze.getCell(x + 1, y));
                solution.add("added " + maze.getCell(x + 1, y));
                s += "added " + maze.getCell(x+1, y) + "\n";
            }
            if(y < maze.getHeight() - 1 && (maze.getCell(x, y + 1).getValue() == 0 || maze.getCell(x, y + 1).getValue() == 3) && maze.getCell(x,y+1).isVisited() == false){
                q.add(maze.getCell(x, y+1));
                maze.getCell(x, y+1).setPrevCell(maze.getCell(x, y));
                maze.getCell(x, y+1).visit();
                // System.out.println("added " + maze.getCell(x, y+1));
                solution.add("added " + maze.getCell(x, y+1));
                s += "added " + maze.getCell(x, y +1) + "\n";
            }
            q.remove();
        }
        while(path.peek().getPrevCell() != null){
            int x = path.peek().getPrevCell().getX();
            int y = path.peek().getPrevCell().getY();
            path.push(maze.getCell(x, y));
        }
        while(!path.isEmpty()){
            //System.out.println(solution.pop());
            Cell curr = path.pop();
            solution.add("solving " + curr);
            s += "solving " + curr + "\n";
        }
        try{
            int count = 0;
            while(!f.createNewFile()){
                count++;
                f = new File(".\\solutions\\solution" + count + ".txt");
            }
            FileWriter fw = new FileWriter(f);
            fw.write(s);
            fw.close();
        }catch(Exception e){
            System.out.println("error writing to file");
        }
        return solution.toArray(new String[solution.size()]);
    }
}

File: pythonver/bfs_solver.py
from typing import Any
from maze import Maze
from cell import Cell
from collections import deque

class BFS_Solver:

    def __init__(self, maze : Maze):
        self._maze = maze
        self._start = maze.get_start()
        self._finish = None
        self._path = []
        self._steps = []

    def solve(self) -> list[str]:
        q = deque()
        sol = []
        q.append(self._start)
        sol.append('added ' + str(self._start[0]) + ',' + str(self._start[1]))
        while q:
            curr = q.popleft()
            x, y = curr
            curr = self._maze.get_cell(x, y)
            if curr.get_value() == 3:
                self._finish = curr
                break
            elif not curr.get_visited():
                curr.visit()
                sol.append('visited ' + str(curr))
                self._steps.append(curr)
                for cell in self._get_neighbors(curr):
                    sol.append('added ' + str(cell))
                    q.append((cell.get_x(), cell.get_y()))
        
        if not self._finish:
            return None
        else:
            path = self._get_path()
            for cell in path:
                sol.append('solving ' + str(cell))
        return sol
        
    def _get_path(self) -> list[Cell]:
        path = []
        curr = self._finish
        while curr:
            path.append(curr)
            curr = curr.get_prev_cell()
        return path[::-1]

    def _get_neighbors(self, cell : Cell) -> list[Cell]:
        x = cell.get_x()
        y = cell.get_y()
        neighbors = []
        if x > 0:
            curr = self._maze.get_cell(x-1, y)
            if curr.get_value() != 1 and not curr.get_visited():
                neighbors.append(curr)
                curr.set_prev_cell(cell)
        if x < self._maze.get_width() - 1:
            curr = self._maze.get_cell(x+1, y)
            if curr.get_value() != 1 and not curr.get_visited():
                neighbors.append(curr)
                curr.set_prev_cell(cell)
        if y > 0:
            curr = self._maze.get_cell(x, y-1)
            if curr.get_value() != 1 and not curr.get_visited():
                neighbors.append(curr)
                curr.set_prev_cell(cell)
        if y < self._maze.get_height() - 1:
            curr = self._maze.get_cell(x, y+1)
            if curr.get_value() != 1 and not curr.get_visited():
                neighbors.append(curr)
                curr.set_prev_cell(cell)
        return neighbors

